[Concurrency] Introduce TaskOptionRecord for spawn options

ktoso opened the pull request:
Work in progress on adding `TaskOptionsRecord` to all task creating and waiting.

One such record can be "the executor to use for the spawn or resume" but we can add various new ones in the future.

Currently doing something incredibly silly somewhere where the signatures don't match up... I'll figure it out soon.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -529,6 +529,73 @@ inline void Job::runInFullyEstablishedContext() {
+
+  FLAGSET_DEFINE_FIELD_ACCESSORS(Kind, Kind_width, TaskOptionRecordKind,
+                                 getKind, setKind)
+};
```
This should go in MetadataValues.h, since the compiler will need to generate it.

ktoso replied:
Ok, will do

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -529,6 +529,73 @@ inline void Job::runInFullyEstablishedContext() {
+
+/// The abstract base class for all options that may be used
+/// to configure a newly spawned task.
+class TaskOptionRecord {
```
The rest of this should go in a new header.  Please forward-declare `TaskOptionRecord` in `Concurrency.h` rather than including that header.

ktoso replied:
Okey!

ktoso committed:
[Concurrency] introduce task options, and change ABI to accept them

introduce new options parameter to all task spawning

[Concurrency] ABI for asynclet start to accept options

[Concurrency] fix unittest usages of changed task creation ABI

[Concurrency] introduce constants for parameter indexes in ownership

[Concurrency] fix test/SILOptimizer/closure_lifetime_fixup_concurrency.swift

ktoso commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -1162,8 +1161,35 @@ void SILGenFunction::emitPatternBinding(PatternBindingDecl *PBD,
+
       alet = emitAsyncLetStart(
           loc,
+          options,
```
Having a lot of trouble figuring this one out...

I'm trying to get a `null` `SILValue` that will "fit" here, the Swift type declaration of what this ends up calling is:

```
@_silgen_name("swift_asyncLet_start")
public func _asyncLetStart<T>(
  asyncLet: Builtin.RawPointer,
  options: Builtin.RawPointer?,
  operation: @Sendable () async throws -> T
)
```

and the RuntimeFunction

```
FUNCTION(AsyncLetStart,
         swift_asyncLet_start, SwiftCC,
         ConcurrencyAvailability,
         RETURNS(VoidTy),
         ARGS(SwiftAsyncLetPtrTy,         // AsyncLet*, alias for Int8PtrTy
              SwiftTaskOptionRecordPtrTy, // options
              TypeMetadataPtrTy,          // futureResultType
              Int8PtrTy,                  // closureEntry
              OpaquePtrTy,                // closureContext
         ),
         ATTRS(NoUnwind, ArgMemOnly))
```

where   SwiftTaskOptionRecordPtrTy is just

```
  SwiftTaskOptionRecordPtrTy = Int8PtrTy; // we pass it opaquely (TaskOptionRecord*)
```

So I need to "make" a `null` of `i8*` in SIL, and get that as a SILValue, we're all happy then...

The various attempts are commented here... in general It seems an unchecked cast of a 0 literal would be right but that ends up in a nasty crash now

```

0.	Program arguments: /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift-frontend -frontend -c -primary-file /Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift -target x86_64-apple-macosx10.9 -enable-objc-interop -sdk /Applications/Xcode-Internal.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -F /Applications/Xcode-Internal.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -F /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/lib -module-cache-path /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache -swift-version 4 -define-availability "SwiftStdlib 5.5:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" -enable-experimental-concurrency -target-sdk-version 11.3 -parse-as-library -module-name main -o /var/folders/w1/hmg_v8p532d800g08jtqtddc0000gn/T/async_task_cancellation_early-dd56ee.o
1.	Swift version 5.5-dev (LLVM b12ac9fa38462e7, Swift 038af80a880ce85)
2.	Compiling with effective version 4.1.50
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "/Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift")
4.	While silgen emitFunction SIL function "@$s4main30test_detach_cancel_child_earlyyyYaF".
 for 'test_detach_cancel_child_early()' (at /Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift:14:1)
5.	While silgen closureexpr SIL function "@$s4main30test_detach_cancel_child_earlyyyYaFSbyYaYbcfU_".
 for expression at [/Users/ktoso/code/swift-project/swift/test/Concurrency/Runtime/async_task_cancellation_early.swift:16:44 - line:27:3] RangeText="{
    async let childCancelled: Bool = { () -> Bool in
      await Task.sleep(2_000_000_000)
      return Task.isCancelled
    }()

    let xx = await childCancelled
    print("child, cancelled: \(xx)") // CHECK: child, cancelled: true
    let cancelled =  Task.isCancelled
    print("self, cancelled: \(cancelled )") // CHECK: self, cancelled: true
    return cancelled
  "
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000109620a67 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x000000010961fc15 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x0000000109621096 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007fff204b7d7d _sigtramp + 29
4  libsystem_platform.dylib 0x00007ffeeb4c8c90 _sigtramp + 18446744072820428592
5  swift-frontend           0x00000001049b9383 swift::SILBuilder::createIntegerLiteral(swift::IntegerLiteralExpr*) + 131
6  swift-frontend           0x0000000104f07213 swift::Lowering::SILGenFunction::emitPatternBinding(swift::PatternBindingDecl*, unsigned int) + 963
7  swift-frontend           0x0000000104f0769d swift::Lowering::SILGenFunction::visitPatternBindingDecl(swift::PatternBindingDecl*) + 45
8  swift-frontend           0x0000000104f82e1d swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 429
9  swift-frontend           0x0000000104f82c65 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
10 swift-frontend           0x0000000104f3599e swift::Lowering::SILGenFunction::emitClosure(swift::AbstractClosureExpr*) + 430
11 swift-frontend           0x0000000104eb4a9a swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 1706
12 swift-frontend           0x0000000104eb7c16 swift::Lowering::SILGenModule::emitClosure(swift::AbstractClosureExpr*) + 134
13 swift-frontend           0x0000000104f2bb3c (anonymous namespace)::RValueEmitter::visitAbstractClosureExpr(swift::AbstractClosureExpr*, swift::Lowering::SGFContext) + 124
14 swift-frontend           0x0000000104f1f730 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 128
15 swift-frontend           0x0000000104f16357 swift::Lowering::SILGenFunction::emitRValueAsSingleValue(swift::Expr*, swift::Lowering::SGFContext) + 55
16 swift-frontend           0x0000000104f254ab swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 24059
17 swift-frontend           0x0000000104f16357 swift::Lowering::SILGenFunction::emitRValueAsSingleValue(swift::Expr*, swift::Lowering::SGFContext) + 55
18 swift-frontend           0x0000000104e9d2c0 swift::Lowering::ArgumentSource::getAsSingleValue(swift::Lowering::SILGenFunction&, swift::Lowering::SGFContext) && + 352
19 swift-frontend           0x0000000104f03405 swift::Lowering::SILGenFunction::emitConvertedRValue(swift::SILLocation, swift::Lowering::Conversion const&, swift::Lowering::SGFContext, llvm::function_ref<swift::Lowering::ManagedValue (swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::SGFContext)>) + 357
20 swift-frontend           0x0000000104e9d438 swift::Lowering::ArgumentSource::getConverted(swift::Lowering::SILGenFunction&, swift::Lowering::Conversion const&, swift::Lowering::SGFContext) && + 120
21 swift-frontend           0x0000000104ed4cbe (anonymous namespace)::ArgEmitter::emit(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 2686
22 swift-frontend           0x0000000104ec3177 (anonymous namespace)::ArgEmitter::emitSingleArg(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 679
23 swift-frontend           0x0000000104ed3de3 (anonymous namespace)::ArgEmitter::emitPreparedArgs(swift::Lowering::PreparedArguments&&, swift::Lowering::AbstractionPattern) + 147
24 swift-frontend           0x0000000104edd2de (anonymous namespace)::CallSite::emit(swift::Lowering::SILGenFunction&, swift::Lowering::AbstractionPattern, swift::CanTypeWrapper<swift::SILFunctionType>, (anonymous namespace)::ParamLowering&, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::SmallVectorImpl<(anonymous namespace)::DelayedArgument>&, swift::Lowering::CalleeTypeInfo::ForeignInfo const&) && + 926
25 swift-frontend           0x0000000104edca6f (anonymous namespace)::CallEmission::emitArgumentsForNormalApply(swift::Lowering::AbstractionPattern, swift::CanTypeWrapper<swift::SILFunctionType>, swift::Lowering::CalleeTypeInfo::ForeignInfo const&, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::Optional<swift::SILLocation>&) + 1631
26 swift-frontend           0x0000000104ec744a (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext) + 2794
27 swift-frontend           0x0000000104ec4843 swift::Lowering::SILGenFunction::emitApplyExpr(swift::ApplyExpr*, swift::Lowering::SGFContext) + 51
28 swift-frontend           0x0000000104f1f716 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 102
29 swift-frontend           0x0000000104f13b62 swift::Lowering::SILGenFunction::emitExprInto(swift::Expr*, swift::Lowering::Initialization*, llvm::Optional<swift::SILLocation>) + 258
30 swift-frontend           0x0000000104f07474 swift::Lowering::SILGenFunction::emitPatternBinding(swift::PatternBindingDecl*, unsigned int) + 1572
31 swift-frontend           0x0000000104f0769d swift::Lowering::SILGenFunction::visitPatternBindingDecl(swift::PatternBindingDecl*) + 45
32 swift-frontend           0x0000000104f82e1d swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 429
33 swift-frontend           0x0000000104f82c65 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
34 swift-frontend           0x0000000104f3517a swift::Lowering::SILGenFunction::emitFunction(swift::FuncDecl*) + 362
35 swift-frontend           0x0000000104eb5c3e swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 6222
36 swift-frontend           0x0000000104eb7887 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef, bool) + 231
37 swift-frontend           0x0000000104eb43ce swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*) + 126
38 swift-frontend           0x0000000104eb9ea4 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const + 1988
39 swift-frontend           0x0000000104f827c0 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) + 128
40 swift-frontend           0x0000000104ebdcb3 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) + 403
41 swift-frontend           0x0000000104ebae01 swift::performASTLowering(swift::FileUnit&, swift::Lowering::TypeConverter&, swift::SILOptions const&) + 113
42 swift-frontend           0x000000010480a1d2 performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 226
43 swift-frontend           0x00000001047fe7a6 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4838
44 swift-frontend           0x0000000104751212 main + 514
45 libdyld.dylib            0x00007fff2048df3d start + 1
<unknown>:0: error: unable to execute command: Segmentation fault: 11
<unknown>:0: error: compile command failed due to signal 11 (use -v to see invocation)
```

ktoso commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -1162,8 +1161,35 @@ void SILGenFunction::emitPatternBinding(PatternBindingDecl *PBD,
+      CanType OptRawPointerTy =
+          OptionalType::get(RawPointerTy)->getCanonicalType();
+//      SILValue options = getOptionalNoneValue(
+//          loc, getTypeLowering(OptRawPointerTy));
```
this was my first attempt though this results in a `i64 0` value, which then fails the type check because the function expects an `i8*`.

The function: `declare extern_weak swiftcc void @swift_asyncLet_start(i8* %0, i8* %1, %swift.type* %2, i8* %3, %swift.opaque* %4) #5` (where the %1 is what we need to pass the null too)

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test macOS Platform

ktoso commented:
@swift-ci please smoke test Linux Platform

ktoso commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -621,6 +621,7 @@ IRGenModule::IRGenModule(IRGenerator &irgen,
   AsyncFunctionPointerPtrTy = AsyncFunctionPointerTy->getPointerTo(DefaultAS);
   SwiftTaskPtrTy = SwiftTaskTy->getPointerTo(DefaultAS);
   SwiftAsyncLetPtrTy = Int8PtrTy; // we pass it opaquely (AsyncLet*)
+  SwiftTaskOptionRecordPtrTy = Int64Ty; // Builtin.RawPointer? that we get as (TaskOptionRecord*)
```
Why not `Int8PtrTy` here, like with `SwiftAsyncLetPtrTy`?

ktoso replied:
Hah yeah this is what took me a week to figure out weirdly enough... With the kind of some folks we figured out that the lowering of the `Optional<Builtin.RawPointer>` ends up as int64 and we must have those two align. I indeed wanted `i8*` here but there didn't seem to be a way to solve it AFAICS -- I'm sure these is some way tho

DougGregor commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -769,7 +769,7 @@ OperandOwnership
                                                               StringRef attr) {
   // The function operand is consumed by the new task.
-  if (&op == &bi->getOperandRef(2))
+  if (&op == &bi->getOperandRef(3))
```
Hmm. It might be worth creating a named constant for this.

DougGregor commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -783,7 +783,7 @@ OperandOwnership
                                                                    StringRef attr) {
   // The function operand is consumed by the new task.
-  if (&op == &bi->getOperandRef(3))
+  if (&op == &bi->getOperandRef(4))
```
... and this

DougGregor commented on stdlib/public/Concurrency/AsyncLet.swift:
```diff
@@ -20,6 +20,7 @@ import Swift
 @_silgen_name("swift_asyncLet_start")
 public func _asyncLetStart<T>(
   asyncLet: Builtin.RawPointer,
+  options: Builtin.RawPointer?, /// <---- pass nil here
```
I don't think you need this comment now ;)

ktoso replied:
ha, yes thanks :)

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
(macOS CI has a bad day, waiting it out for now)

ktoso commented:
@swift-ci please smoke test macOS Platform

ktoso committed:
[Concurrency] move isAsyncTask into flags of task creation

ktoso commented:
@swift-ci please smoke test

DougGregor commented on include/swift/ABI/TaskOptions.h:
```diff
@@ -0,0 +1,97 @@
+/// executor should be used instead, most often this may mean the global
+/// concurrent executor, or the enclosing actor's executor.
+class ExecutorTaskOptionRecord : public TaskOptionRecord {
+  ExecutorRef *Executor;
```
I'm pretty sure you want `ExecutorRef` rather than `ExecutorRef *` here.

ktoso replied:
Oh, okey, will do

DougGregor replied:
Nah, I got it.

DougGregor reviewed (approved):
Just the one request left, and otherwise LGTM

DougGregor commented:
Force-merging to unblock progress
